authentication/oidc.go

```diff
@@ -45,71 +47,127 @@ type OIDCConfig struct {
 // Middleware is a convenience type for functions that wrap http.Handlers.
 type Middleware func(http.Handler) http.Handler
 
-// NewOIDC creates a single http.Handler and a set of Middlewares for all
+// OIDCHandlers holds handlers and a set of middlewares for all
 // tenants that is able to authenticate requests and provide the
 // authorization code grant flow for users.
-func NewOIDC(
-	logger log.Logger,
-	reg prometheus.Registerer,
-	prefix string,
-	configs []TenantOIDCConfig,
-) (http.Handler, map[string]Middleware, []error) {
-	handlers := map[string]http.Handler{}
-	middlewares := map[string]Middleware{}
-	warnings := make([]error, 0, len(configs))
-
-	tenantErrors := prometheus.NewGauge(prometheus.GaugeOpts{
+type OIDCHandlers struct {
+	mtx         sync.RWMutex
+	handlers    map[string]http.Handler
+	middlewares map[string]Middleware
+
+	logger     log.Logger
+	retryGauge prometheus.Gauge
+}
+
+// NewOIDCHandlers instantiates OIDC handlers.
+func NewOIDCHandlers(l log.Logger, reg prometheus.Registerer) *OIDCHandlers {
+	tenantsFailing := prometheus.NewGauge(prometheus.GaugeOpts{
 		Namespace: "observatorium",
 		Subsystem: "api",
-		Name:      "tenant_registration_failures",
-		Help:      "The number of tenants for which OIDC provider instantiation failed.",
+		Name:      "tenants_failing_registrations",
+		Help:      "The number of tenants for which OIDC provider instantiation is failing.",
 	})
 
-	reg.MustRegister(tenantErrors)
+	reg.MustRegister(tenantsFailing)
 
-	for _, c := range configs {
-		p, err := NewProvider(context.TODO(), logger, getCookieForTenant(c.Tenant), "/"+c.Tenant, c.OIDCConfig)
-		if err != nil {
-			warnings = append(warnings, fmt.Errorf("failed to instantiate OIDC provider for tenant %q: %w", c.Tenant, err))
+	return &OIDCHandlers{
+		handlers:    make(map[string]http.Handler),
+		middlewares: make(map[string]Middleware),
+		logger:      l,
+		retryGauge:  tenantsFailing,
+	}
+}
 
-			tenantErrors.Inc()
+// AddOIDCForTenant adds a new handler and middleware for a new tenant, based on
+// the tenant OIDC configuration. If a connection with the OIDC provider
+// cannot be established, the method will retry to connect with an increasing
+// backoff.
+func (oh *OIDCHandlers) AddOIDCForTenant(prefix string, config TenantOIDCConfig) {
+	const (
+		loginRoute    = "/login"
+		callbackRoute = "/callback"
+	)
+
+	ctx := context.Background()
+	b := backoff.New(ctx, backoff.Config{
+		Min:        500 * time.Millisecond,
+		Max:        5 * time.Second,
+		MaxRetries: 0, // retry indefinitely
+	})
 
-			continue
-		}
+	go func() {
+		var isFailing bool
+		for b.Reset(); b.Ongoing(); {
+			p, err := NewProvider(ctx, oh.logger, getCookieForTenant(config.Tenant), "/"+config.Tenant, config.OIDCConfig)
+			if err != nil {
+				level.Warn(oh.logger).Log("msg", "failed to instantiate OIDC provider for tenant", "tenant", config.Tenant, "error", err)
 
-		r := chi.NewRouter()
+				if !isFailing {
+					oh.retryGauge.Inc()
```

Comment:
I would argue that retry counter could be better here. Few reasons:

* Overall prefer counters over gauges. Counters are useful when service is restarting/crashing. Gauges not.
* We need to maintain extra state using `isFailing`, making code bit more complex
*  We have similar info, if not better -> how many retries we did. Becuase what if gauge is 1 but we got stuck on second retry forever because of sum bug, and we never retry. We don't have that info with gague.

Reply:
I see your points, as originally I used counter, but I really thought @onprem's idea with gauge was neater. There is some marginal complexity increase in keeping the state, but conversely to your points, gauge enables us to see how many tenants are failing and for how long, which I think is the crucial data point here. Recording just the number of retries won't directly tell us how widespread the issue is.

@onprem any thoughts?

Reply:
We briefly touched upon this on the community call, it seems the general agreement in the end was to keep this simple for now, whether we do gauge or counter, and see how our needs evolve in the future, e.g. with more tenants.

Reply:
> us to see how many tenants are failing and for how long, which I think is the crucial data point here. Recording just the number of retries won't directly tell us how widespread the issue is.

You can do same with counter, no? (: But if on the community call we are happy with counter, let's go 💪🏽

Reply:
I guess we were happy with either :sweat_smile: Let's keep it and come back to this if we feel it needs improving in the future.